### PR TITLE
Binding/xfconf

### PR DESCRIPTION
**Pull request by Eiskasten**
## Basics

- [ ] Short descriptions of your changes are in the release notes
      (added as entry in `doc/news/_preparation_next_release.md` which
      contains `_(my name)_`)
      **Please always add something to the release notes.**
- [ ] Details of what you changed are in commit messages
      (first line should have `module: short statement` syntax)
- [ ] References to issues, e.g. `close #X`, are in the commit messages.
- [ ] The buildservers are happy. If not, fix **in this order**:
  - [ ] add a line in `doc/news/_preparation_next_release.md`
  - [ ] reformat the code with `scripts/dev/reformat-all`
  - [ ] make all unit tests pass
  - [ ] fix all memleaks
- [ ] The PR is rebased with current master.

## Checklist

- [ ] I added unit tests for my code
- [ ] I fully described what my PR does in the documentation
      (not in the PR description)
- [ ] I fixed all affected documentation
- [ ] I added code comments, logging, and assertions as appropriate (see [Coding Guidelines](https://master.libelektra.org/doc/CODING.md))
- [ ] I updated all meta data (e.g. README.md of plugins and [METADATA.ini](https://master.libelektra.org/doc/METADATA.ini))
- [ ] I mentioned [every code](/.reuse/dep5) not directly written by me in [reuse syntax](https://reuse.software/)

## Review

- [ ] Documentation is introductory, concise, good to read and describes everything what the PR does
- [ ] Examples are well chosen and understandable
- [ ] Code is conforming to [our Coding Guidelines](https://master.libelektra.org/doc/CODING.md)
- [ ] APIs are conforming to [our Design Guidelines](https://master.libelektra.org/doc/DESIGN.md)
- [ ] Code is consistent to [our Design Decisions](https://master.libelektra.org/doc/decisions)

## Labels

- [ ] Add the "work in progress" label if you do not want the PR to be reviewed yet.
- [ ] Add the "ready to merge" label **if the basics are fulfilled** and no further pushes are planned by you.

### CODE REVIEW

**Comment by markus2330** [comment]
Great to see this!

Please do not forget issues/PRs concerning new-backend :rocket:

**Comment by Eiskasten** [comment]
@kodebach Do you know why the jenkins CI fails in this case? What does the jenkins state "unstable" mean? I cannot find any failing jobs or tests.

**Comment by kodebach** [comment]
No idea, @lukashartl or @markus2330 might know, but I'd just rerun the pipeline and see if the problem goes away.

**Comment by Eiskasten** [comment]
Ok, I will try that

**Comment by Eiskasten** [comment]
jenkins build libelektra please

**Comment by Eiskasten** [comment]
jenkins build libelektra please

**Comment by Eiskasten** [comment]
jenkins build libelektra please

**Comment by Eiskasten** [comment]
jenkins build libelektra please

**Comment by Eiskasten** [comment]
jenkins build libelektra please

**Comment by Eiskasten** [comment]
jenkins build libelektra please

**Comment by Eiskasten** [comment]
@lukashartl currently the jenkins CI fails for this PR. Interestingly, when I trigger a rebuild for the same commit, the failure was triggered by a different job. You can see that by clicking the arrows on top left and compare the commit checksums. Do you have an idea what can cause this? Are other PRs affected as well?

**Comment by mpranj** [comment]
@Eiskasten no idea what causes this. A silly way that could trigger a proper rebuild: rebase from master and force push.

We once had a very rare circumstance where we had to close the PR and reopen a new one, but that's a last resort.

**Comment by Eiskasten** [comment]
jenkins build libelektra please

**Comment by Eiskasten** [comment]
@mpranj Thank you for your advise, it works now. However, it still fails randomly sometimes.

**Comment by Eiskasten** [comment]
@markus2330 @tucek I know the review week did not start yet, but testing the Xfce session might not be straight forward. Please just follow the `Using the binding as a replacement in Xfce` section in the `src/bindings/xfconf/README.md` and let me know if you have problems.

**Comment by tucek** [comment]
@Eiskasten the build does not finish, please address these issues. i willl have a look at the changes in the meanwhile

**Comment by tucek** [comment]
@mpranj why does this link chekc fail?
```
if test -s broken_links.txt; then
  printf >&2 'Broken Links:\n'
  printf >&2 '—————————————\n'
  cat >&2 broken_links.txt
  printf >&2 '—————————————\n'
  false
fi
Broken Links:
—————————————
src/plugins/crypto/README.md:64:0 https://gpgtools.org
```

**Comment by tucek** [comment]
@Eiskasten please rebase branch to current master

**Comment by kodebach** [comment]
> why does this link check fail?

If a link works locally for you, always attempt re-running the CI job (with the button on GitHub). Link checker issues are often temporary.

**Comment by Eiskasten** [comment]
Done, the failing macOS builds should not affect anything. Building these bindings is disabled on macOS.

File: src/bindings/xfconf/README.md
Comment (tucek) [inline src/bindings/xfconf/README.md:26]: @Eiskasten did you really want to include unfinished TODOs in this PR? (also see below)

**Comment by tucek** [comment]
LGTM, could build and install it, also read the documentation and found it useful. Since this is a C implementation, i would like someone else to have a second look.

File: src/bindings/xfconf/README.md
Comment (tucek) [inline src/bindings/xfconf/README.md:70]: @Eiskasten you mean `xfconf` instead of `xfconfbinding`, right?

Comment (kodebach) [reply src/bindings/xfconf/README.md:70]: I suppose this was the intention:

```suggestion
Build elektra as instructed in the [COMPILE](../../../doc/COMPILE.md#developer-options) and do not forget to include the `xfconf` binding and set the `KDB_DB_SYSTEM` and `CMAKE_INSTALL_PREFIX` paths properly.
```

File: src/bindings/xfconf/CMakeLists.txt
Comment (kodebach) [inline src/bindings/xfconf/CMakeLists.txt:4]: Since you require a specific version, is this version actually checked anywhere?

File: src/bindings/xfconf/README.md
Comment (kodebach) [inline src/bindings/xfconf/README.md:51]: Would be nice, if you could provide that as a shell script...

File: src/bindings/xfconf/elektra-xfconf.c
Comment (kodebach) [inline src/bindings/xfconf/elektra-xfconf.c:5]: Are these global variables really necessary? Is this definitely only executed in a single-threaded context? Otherwise you'll need some thread synchronization...

File: src/bindings/xfconf/elektra-xfconf.c
Comment (kodebach) [inline src/bindings/xfconf/elektra-xfconf.c:58]: I'm not sure what you're trying to do here, but you should probably use unescaped names (or the `key*Name*` APIs). You're not handling any of the various escapes that might appear in the escaped name.

**Review by kodebach — COMMENTED** [review]
I didn't try to run it and only had a quick look a the code, but based on that the PR looks mostly fine to me.

The only issue I have is that you seem to do a lot of key name manipulation with the escaped names. Since you don't handle the escapes properly that could go wrong. However, maybe because of the interaction with `xfconf` the escape stuff is irrelevant anyway. For example, if `xfconf` only allows something like `[a-zA-Z0-9_-]`  your code should be fine, but I don't know if that's actually the case.

File: src/bindings/xfconf/README.md
Comment (tucek) [inline src/bindings/xfconf/README.md:33]: Had to upgrade to Ubuntu 22.04 because of the xfconf 4.16 requirement.
```
sudo apt-get install xfconf glib-2.0 libxfconf-0-dev
```
Without `libxfconf-0-dev` i would still get:
```
-- Could NOT find Xfconf (missing: XFCONF_LIBRARY XFCONF_INCLUDE_DIR)
-- Exclude binding xfconf because xfconf >= 4.16 required for xfconf backend
```

@Eiskasten maybe include this for Ubuntu users?

Comment (tucek) [reply src/bindings/xfconf/README.md:33]: And of course had to enable `xfconf` in bindings and plugins

**Comment by tucek** [comment]
LGTM, after some trougbles i could build and install it, also read the documentation and found it useful, but suggested some improvements. Since this is a C implementation, i would like someone else to have a second look. @kodebach thx

**Comment by Eiskasten** [comment]
> I didn't try to run it and only had a quick look a the code, but based on that the PR looks mostly fine to me.
> 
> The only issue I have is that you seem to do a lot of key name manipulation with the escaped names. Since you don't handle the escapes properly that could go wrong. However, maybe because of the interaction with `xfconf` the escape stuff is irrelevant anyway. For example, if `xfconf` only allows something like `[a-zA-Z0-9_-]` your code should be fine, but I don't know if that's actually the case.

This is according to the official docs: https://docs.xfce.org/xfce/xfconf/start#introduction indeed the case

File: src/bindings/xfconf/CMakeLists.txt
Comment (Eiskasten) [reply src/bindings/xfconf/CMakeLists.txt:4]: No, it was only the version I started with. To be honest, I did not check if there is actually something which requires at least version 4.16. Since 4.16. was released back in december 2020, I thought that this would be sufficient.

File: src/bindings/xfconf/README.md
Comment (Eiskasten) [reply src/bindings/xfconf/README.md:33]: Ok, so the dev package is the same as in Debian. I will include that.

**Comment by Eiskasten** [comment]
Besides the tutorial I think this could be merged. @markus2330 as we discussed in the meeting I would open a separate issue regarding the tutorial.

File: src/bindings/xfconf/CMakeLists.txt
Comment (markus2330) [reply src/bindings/xfconf/CMakeLists.txt:4]: Even Debian stable has 4.16.0-2 so it should be fine.

File: src/bindings/xfconf/CMakeLists.txt
Comment (markus2330) [inline src/bindings/xfconf/CMakeLists.txt:23]: Do we really need this? Is it possible to only build this directory with CMake?

File: src/bindings/xfconf/README.md
Comment (markus2330) [inline src/bindings/xfconf/README.md:13]: As the binding is probably the more important artifact, I would prefer if here is the general information and the plugin links here.

File: src/bindings/xfconf/README.md
Comment (markus2330) [reply src/bindings/xfconf/README.md:24]: tbd

File: src/bindings/xfconf/README.md
Comment (markus2330) [reply src/bindings/xfconf/README.md:49]: I don't know what "place them into you system" means, please rephrase.

File: doc/news/_preparation_next_release.md
Comment (markus2330) [inline doc/news/_preparation_next_release.md:288]: Please describe more what it does etc. (Suitable for later highlight).

File: src/bindings/xfconf/README.md
Comment (markus2330) [inline src/bindings/xfconf/README.md:51]: The shell script does something quite different (much less)?

File: src/bindings/xfconf/README.md
Comment (markus2330) [inline src/bindings/xfconf/README.md:54]: Probably you should describe that first then or first give an introduction of options (and describe pros/cons).

File: src/bindings/xfconf/README.md
Comment (markus2330) [inline src/bindings/xfconf/README.md:61]: Is this still the case?

File: src/bindings/xfconf/README.md
Comment (markus2330) [inline src/bindings/xfconf/README.md:96]:
```suggestion
Note that this binding only handles configuration directly handled within Xfconf.
```

File: src/bindings/xfconf/README.md
Comment (markus2330) [inline src/bindings/xfconf/README.md:98]:
```suggestion
However, since settings such as the Gtk theme are part of the Xfce desktop itself, this binding is not able to configure all aspects of Xfce.
```

File: src/bindings/xfconf/elektra-xfconf-binding.c
Comment (markus2330) [inline src/bindings/xfconf/elektra-xfconf-binding.c:42]: For nicer code formatting:

```suggestion
	require_binding_write_lock ();
	binding->id = last_id++;
```

File: src/bindings/xfconf/elektra-xfconf-channel.c
Comment (markus2330) [inline src/bindings/xfconf/elektra-xfconf-channel.c:354]:
```suggestion
 * @retval 0 on success and 1 otherwise
```

File: src/bindings/xfconf/elektra-xfconf-channel.c
Comment (markus2330) [inline src/bindings/xfconf/elektra-xfconf-channel.c:386]: elektraStrCaseCmp? Or is this really the definition of bool in xfconf?

File: src/bindings/xfconf/elektra-xfconf.c
Comment (markus2330) [reply src/bindings/xfconf/elektra-xfconf.c:85]: Please add this as comment here.

File: src/bindings/xfconf/elektra-xfconf.h
Comment (markus2330) [inline src/bindings/xfconf/elektra-xfconf.h:9]: Should be according guidelines in doc/tutorials/application-integration.md

File: src/bindings/xfconf/scripts/replace-system-xfconf.sh
Comment (markus2330) [inline src/bindings/xfconf/scripts/replace-system-xfconf.sh:2]: Please add very short description (meta-data) what this script does. Move it to the scripts folder and install it, so that it is also available via `kdb replace-system-xfconf`.

File: src/bindings/xfconf/CMakeLists.txt
Comment (markus2330) [inline src/bindings/xfconf/CMakeLists.txt:19]: Please also install the library in its own component (e.g. see src/bindings/swig/lua and scripts/cmake/ElektraPackaging.cmake).

File: src/bindings/xfconf/CMakeLists.txt
Comment (markus2330) [inline src/bindings/xfconf/CMakeLists.txt:20]: With ELEKTRA_UNUSED you should be able to avoid the second parameter?

**Review by markus2330 — CHANGES_REQUESTED** [review]
Looking forward to see this merged! Please ensure proper packaging (so that I can easily test it once it gets merged) and update documentation a bit.

File: src/bindings/xfconf/README.md
Comment (Eiskasten) [reply src/bindings/xfconf/README.md]: Should be done by 6506d0e23efa25047381ef0cd2afa5aed13da3f7 or before

File: doc/news/_preparation_next_release.md
Comment (Eiskasten) [reply doc/news/_preparation_next_release.md]: Should be done by 6506d0e23efa25047381ef0cd2afa5aed13da3f7 or before

File: src/bindings/xfconf/README.md
Comment (Eiskasten) [reply src/bindings/xfconf/README.md]: Should be done by 6506d0e23efa25047381ef0cd2afa5aed13da3f7 or before. I updated the section, I hope it is much more clear now.

File: src/bindings/xfconf/README.md
Comment (Eiskasten) [reply src/bindings/xfconf/README.md]: I just mentioned the new packaged script new. Should be done by 6506d0e23efa25047381ef0cd2afa5aed13da3f7 or before.

File: src/bindings/xfconf/README.md
Comment (Eiskasten) [reply src/bindings/xfconf/README.md]: No, configuring Xfconf properties should be fully possible. Should be updated by 6506d0e23efa25047381ef0cd2afa5aed13da3f7 or before.

File: src/bindings/xfconf/README.md
Comment (Eiskasten) [reply src/bindings/xfconf/README.md]: Should be done by 6506d0e23efa25047381ef0cd2afa5aed13da3f7 or before

File: src/bindings/xfconf/README.md
Comment (Eiskasten) [reply src/bindings/xfconf/README.md]: Should be done by 6506d0e23efa25047381ef0cd2afa5aed13da3f7 or before

File: src/bindings/xfconf/elektra-xfconf-binding.c
Comment (Eiskasten) [reply src/bindings/xfconf/elektra-xfconf-binding.c]: Should be done by 6506d0e23efa25047381ef0cd2afa5aed13da3f7 or before

File: src/bindings/xfconf/elektra-xfconf-channel.c
Comment (Eiskasten) [reply src/bindings/xfconf/elektra-xfconf-channel.c]: Should be done by 6506d0e23efa25047381ef0cd2afa5aed13da3f7 or before

File: src/bindings/xfconf/elektra-xfconf.h
Comment (Eiskasten) [reply src/bindings/xfconf/elektra-xfconf.h:10]: I will have a look on that. I hope that will not change that much.

File: src/bindings/xfconf/scripts/replace-system-xfconf.sh
Comment (Eiskasten) [reply src/bindings/xfconf/scripts/replace-system-xfconf.sh]: Should be done by 6506d0e23efa25047381ef0cd2afa5aed13da3f7 or before

File: src/bindings/xfconf/CMakeLists.txt
Comment (Eiskasten) [reply src/bindings/xfconf/CMakeLists.txt]: I will have a look on that.

**Comment by Eiskasten** [comment]
> Looking forward to see this merged! Please ensure proper packaging (so that I can easily test it once it gets merged) and update documentation a bit.

Thank you for the review. The most points should be resolved. However, a few of them are still open. Hopefully I am able to fix them soon.

File: src/bindings/xfconf/elektra-xfconf.c
Comment (Eiskasten) [reply src/bindings/xfconf/elektra-xfconf.c:89]: Done with b8c304e21c2aa120e38930bf81ef1c377f96d563.

File: src/bindings/xfconf/elektra-xfconf-channel.c
Comment (Eiskasten) [reply src/bindings/xfconf/elektra-xfconf-channel.c]: Done with 35b3c5a02291276ff9108da08cd0b2d9525b101a.

File: src/bindings/xfconf/CMakeLists.txt
Comment (Eiskasten) [reply src/bindings/xfconf/CMakeLists.txt]: Should be done with 4376d1a09a80212e4768477d032a2b1b68595e19.

**Comment by Eiskasten** [comment]
Everything should be corrected as you requested now.

**Pull request merged**
